bzzeth/bzzeth.go

```diff
@@ -99,6 +119,215 @@ func (b *BzzEth) handleMsgFromSwarmNode(p *Peer) func(context.Context, interface
+// this is used in testing .. otherwise it just logs trace
+var finishDeliveryFunc = finishDelivery
+
+func finishDelivery(hashes map[string]bool) {
+	for addr := range hashes {
+		log.Trace("Header delivered", "Address", addr)
+	}
+}
+
+// handleBlockHeaders handles block headers message
+func (b *BzzEth) handleBlockHeaders(ctx context.Context, p *Peer, msg *BlockHeaders) {
+	p.logger.Debug("bzzeth.handleBlockHeaders", "id", msg.ID)
+
+	// retrieve the request for this id
+	req, ok := p.requests.get(msg.ID)
+	if !ok {
+		p.logger.Warn("bzzeth.handleBlockHeaders: nonexisting request id", "id", msg.ID)
+		p.Drop()
+		return
+	}
+
+	// convert rlp.RawValue to bytes
+	var headers [][]byte
+	for _, h := range msg.Headers {
+		headers = append(headers, h)
+	}
+
+	err := b.deliverAndStoreAll(ctx, req, headers)
+	if err != nil {
+		p.logger.Warn("bzzeth.handleBlockHeaders: fatal dropping peer", "id", msg.ID, "err", err)
+		p.Drop()
+	}
+}
+
+// Validates and headers asynchronously and stores the valid chunks in one go
+func (b *BzzEth) deliverAndStoreAll(ctx context.Context, req *request, headers [][]byte) error {
+	errC := make(chan error, len(headers))
+
+	chunks := make([]chunk.Chunk, 0)
+	var chunkL sync.RWMutex
+	var wg sync.WaitGroup
+	for _, h := range headers {
+		hdr := make([]byte, len(h))
+		copy(hdr, h)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ch, err := b.validateHeader(ctx, hdr, req)
+			if err != nil {
+				errC <- err
+				return
+			}
+			chunkL.Lock()
+			defer chunkL.Unlock()
+			chunks = append(chunks, ch)
+		}()
+	}
+
+	// wait for all validations to get over and close the channels
+	wg.Wait()
+	close(errC)
+
+	// Store all the valid header chunks in one shot
+	results, err := b.netStore.Put(ctx, chunk.ModePutUpload, chunks...)
+	if err != nil {
+		for i := range results {
+			ch := chunks[i]
+			log.Warn("bzzeth.store", "hash", ch.Address().Hex(), "err", err)
+			// ignore all other errors, but invalid chunk incurs peer drop
+			if err == chunk.ErrChunkInvalid {
+				return err
+			}
+		}
+	}
+	log.Debug("Stored all headers ", "count", len(chunks))
+
+	// finish storage is used mostly in testing
+	// in normal scenario.. it just logs Trace
+	finishStorageFunc(chunks)
+	return <-errC
+}
+
+// validateHeader check for correctness and validity of the header
+// this also informs the delivery channel about the received header
+func (b *BzzEth) validateHeader(ctx context.Context, header []byte, req *request) (chunk.Chunk, error) {
+	ch := newChunk(header)
+	headerAlreadyReceived, expected := isHeaderExpected(req, ch.Address().Hex())
+	if expected {
+		if headerAlreadyReceived {
+			// header already received
+			return nil, errDuplicateHeader
+		} else {
+			setHeaderAsReceived(req, ch.Address().Hex())
+			req.c <- ch.Address()
+			return ch, nil
+		}
+	} else {
+		// header is not present in the request hash.
+		return nil, errUnsolicitedHeader
+	}
+
+	// TODO: Ethereum block header validation should come here
```

Comment:
probably can be safely removed

Reply:
This is required for phase 2... 

Removed anyway